Make PDAL Locale safe

jjimenezshaw opened the pull request:
Fixes #3450 

## Description of the problem
Many C/C++ functions like `atof`, `strtod`, `to_string`, `stringstream`, etc do take into consideration the C locale (in both directions: number to string and string to number). The C locale is a global setting that can be set with a function like `std::locale::global(std::locale("de_DE.utf8"));` (in this example for the German locale... that has to be installed in the OS at execution time). https://en.cppreference.com/w/cpp/locale/locale/global

The global locale can be set by any application or library, and affects globally the running program, including any linked library.

Some locales (like German, Spanish, French, Dutch, etc) have a different number format, using `,` (comma) as decimal separator, and `.` (period) as thousand separator. (Yes, the opposite as English). Not all similar locales behave the same way. For instance `de_DE` uses `,` as decimal separator, while `de_CH` (Swiss German) uses `.`

## Simple example
```C++
#include <locale>
#include <iostream>
#include <sstream>
int main()
{
    std::locale::global(std::locale("de_DE.utf8"));
    float a = atof("42.5");
    std::cout << a << std::endl;
    std::ostringstream oss;
    oss << 3.14 << " " << 65535;
    std::cout << oss.str() << std::endl;
    std::cout << std::to_string(1.23456789) << std::endl;
    std::istringstream iss("2.5");
    double d = 0;
    iss >> d;
    std::cout << d << std::endl;
    return 0;
}
```
produces this output (remember, you have to install that locale in your system)
```
42
3,14 65.535
1,234568
25
```
instead of what you may expect
```
42.5
3.14 65535
1.234568
2.5
```
## Consequences in PDAL
PDAL is using JSON format for many string parsing. That means that numbers MUST use `.` as decimal separator, regardless the locale set by anybody/anything. Currently the code is not safe against a change in the global locale, or even calling `set::locale`.

## Solution used
In C++, you can "imbue" a locale into a stream to modify the behavior of that single instance. Setting `std::locale::classic()` sets the standard US English behavior. Actually it was (only) already done in PDAL in the file `io/PcdHeader.cpp`. To avoid calling `imbue(std::locale::classic())` everywhere, a wrapper class `ClassicLocaleStream` is created that calls `imbue` in its constructor.

In this PR the call to `imbue` only appears twice: in `pdal/util/Utils.hpp` (where other functions to convert to and from string were) and in `plugins/e57/libE57Format/src/Common.h` as E57 code has the same problem.

Unfortunately `TileDB` has the same problem (easy to check setting the global locale in their file `test/src/unit.cc`). I did open [an issue in their repository](https://github.com/TileDB-Inc/TileDB/issues/3555).

## Testing
### How to run test with any specific locale
In the first commit I modify `pdal/pdal_test_main.hpp` to understand the environment variable `TEST_SET_LOCALE`, and call `std::locale::global` with that value. That may help you to run the tests in any locale without recompiling.
### Test in GitHub Actions 
To test it I have created a new test run in Linux. In Linux it is easy to install a locale in the OS. This test `Test with German locale` is run _in addition_ to the normal tests (see the second commit).
### Disabling TileDB tests
4 tests for TileDB are currently failing. I guess it is due to the same problem in that library. While it is fixed in their repo, I have disabled them in that new test set.
### What about Windows?
I do not have a Windows environment to test it. And I do not know how to set it up in GitHub Actions. @wrenoud tested the problem some weeks ago and it was reproducible (thanks for your help).

## Comments
### It happens also with integer values
The thousands separator may be used (see simple example above), that will break any future parsing.
### Finding "every" use case in the code
To fix this problem I have run the test in Linux. Many of them where failing. However, if a test is not using numbers that trigger the failure (for instance, floating point numbers without decimal values), it may not be detected in the tests... and not fixed. I have not replaced every `stringstream` but only those that I though were needed. For instance I have not changed the log messages... that may confuse the user showing a number in a format that is not the same as in the file.

jjimenezshaw commented:
> At first glance this strikes me as excessive change. The issue that you described is parsing floating point input. Is there some additional issue that has caused trouble?

The issue is both for input and output. And not only floating point, but also integers, as I explain in the description.

abellgithub commented:
Is the thousand's separator an issue?

On Sun, Oct 9, 2022, 1:10 PM Javier Jimenez Shaw ***@***.***>
wrote:

> At first glance this strikes me as excessive change. The issue that you
> described is parsing floating point input. Is there some additional issue
> that has caused trouble?
>
> The issue is both for input and output. And not only floating point, but
> also integers, as I explain in the description.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PDAL/PDAL/pull/3873#issuecomment-1272586646>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKBMMDMIOY5KPZUXI3EVZLWCL37ZANCNFSM6AAAAAARAUIWBQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

jjimenezshaw commented:
> Is the thousand's separator an issue?

Well, I "suffered" it in a unittest. If you write the string `65.535`, how should it be parsed later? (I use that number in the example in the description).

abellgithub commented:
If you're typing an integer like "12.345", then clearly you're expecting the locale to apply. This is why I find this entire issue so fraught. I don't understand why a user doesn't simply disable the locale interpretation if that's what they want. PDAL CAN'T know what you expect, and when you set your local to German or Spanish or whatever, I would think that you should expect the translation that is specified by the locale. It seems bold of PDAL or any other library to override what the user has requested. I don't believe that PDAL does its own floating point interpretation from strings (though there may some odd case). I believe there are home-grown integer interpretations, but those could be replaced.

abellgithub commented:
I'll add that I believe that locale support in general was a bad thing, but we are where we are. It's only a little bit worse than wstring.

abellgithub commented:
Your initial issue said something like "1e-3" in some context didn't work. Should it have worked? If not why not?

Anyway, I'm not in charge, but I don't much like the idea of throwing the locale away when someone has set it. If 1e-3 is valid floating point in the german locale and wasn't accepted, I think we should examine that issue and deal with it, rather than try to force users into C/Classic locale despite them having set something else.

jjimenezshaw commented:
@abellgithub we already discussed this in https://github.com/PDAL/PDAL/issues/3450

When I said "if you write the string `65.535`" I wanted to say that PDAL is writing `65.535` in some cases in a formatted file. That is not what you want for an integer value, but `65535`. As @hobu said PDAL format should be like JSON, that does not care about the locale of the user. (You can ask your friends living in Europe how crazy is to import some text files in Excel, as @wrenoud said once).

PDAL currently may be writing and reading files/streams taking into consideration the user locale!

This PR tries to _fix_ PDAL to behave as you really expect.

Otherwise, why are that many unit test failing just setting a different locale? It should be "immune", as a proper JSON parser.

> PDAL CAN'T know what you expect, and when you set your local to German or Spanish or whatever, I would think that you should expect the translation that is specified by the locale.

When your library is reading or writing a file/stream in a specified format that says that decimal separator must be `.`, it _must_ ignore the locale of the running computer, and read/write numbers using `.` as decimal separator. Unfortunately C/C++ formatting functions are not doing that, so you have to take special care when your code is reading or writing formatted numbers. (I also hate this problem in C/C++, but "it is what it is". I cannot change the standard).

> Your initial issue said something like "1e-3" in some context didn't work. Should it have worked? If not why not?

The problem there is that PDAL in my German computer was writing somewhere internally the string `0,001` when I provided the C++ value `double(1e-3)`. PDAL was not working in my computer to create a LAS file (but I did for other colleagues) even when I was not providing any text string. That's how I found the problem initially.

jjimenezshaw commented:
https://github.com/TileDB-Inc/TileDB/issues/3555#issuecomment-1275469201

rouault commented on pdal/XMLSchema.cpp:
```diff
@@ -431,7 +431,9 @@ bool XMLSchema::load(xmlDocPtr doc)
-                dim.m_min = std::atof((const char*)n);
+                Utils::ClassicLocaleIStringStream iss((const char*)n);
+                iss >> dim.m_min;
+                iss.str("");
```
is ``iss.str("");`` really needed ? If so, a comment in the code would be welcome. Same at line 449

jjimenezshaw replied:
Thanks @rouault . I was afraid that the destructor of `istringstream` could conflict with `xmlFree` that is called before the constructor. Now reading carefully the documentation of `istringstream`, I see that it is created with a copy of `n`, so it is safe (if I understood correctly). I will remove that line.

rouault commented:
Looks good to me. We had the same sort of issues in GDAL and PROJ that were solved by using std::locale::classic() at relevant places.

jjimenezshaw commented:
> Please rebase current master against this

Already rebased. All actions are green.

hobu reviewed (changes requested):
I'll admit I'm anxious about this. I recognize the need, and this is going in the correct direction, but I want to see a bit more in two areas:

1) There's not enough tests that have been added. ClassicLocaleStringStream and friends should have some explicit tests that demonstrate behavior and usage. Also, the place where locale stuff is the most difficult to deal with, Windows, is not tested at all.

2) There's some spots where creating big fat streams all the time are going to hurt performance. I tried to identify a few of those, but there might be some more. Updating them probably means changing PDAL's code a bit more.

rouault commented:
Maybe the following can help
```
#include <string>
#include <locale>
#include <sstream>

double c_locale_stod(const std::string& s)
{
    std::istringstream iss(s);
    iss.imbue(std::locale::classic());
    double d;
    iss >> d;
    if (!iss.eof() || iss.fail()) {
        throw std::invalid_argument("non double value");
    }
    return d;
}


double c_locale_stod_thread_local(const std::string& s)
{
    static thread_local std::istringstream iss = []() {
        std::istringstream l_iss;
        l_iss.imbue(std::locale::classic());
        return l_iss;
    }();
    double d;
    iss.clear();
    iss.str(s);
    iss >> d;
    if (!iss.eof() || iss.fail()) {
        throw std::invalid_argument("non double value");
    }
    return d;
}

int main() {
    std::string s("1.23");
    for(int i = 0; i < 1000 * 1000; ++i)
    {
        //std::stod(s);
        //c_locale_stod_thread_local(s);
        //locale_stod(s);
    }
    return 0;
}
```
On my Linux box, above test code when built with -O2:
when using std::stdod(), timing is 77 ms
when using c_locale_stod_thread_local(), timing is 195 ms
when using c_locale_stod(), timing is 487 ms

jjimenezshaw commented on plugins/e57/libE57Format/src/CheckedFile.cpp:
```diff
@@ -276,14 +276,14 @@ CheckedFile& CheckedFile::operator<<(const ustring& s)
 CheckedFile& CheckedFile::operator<<(int64_t i)
 {
-   stringstream ss;
+   ClassicLocaleStream<stringstream> ss;
```
Thanks @hobu I hope I addressed all your points.

> * There's not enough tests that have been added. ClassicLocaleStringStream and friends should have some explicit tests that demonstrate behavior and usage. Also, the place where locale stuff is the most difficult to deal with, Windows, is not tested at all.
> 

I added tests for ClassicLocale stringstreams. About Windows, unfortunately I do not have a computer to test, and neither the knowledge to configure the GitHub actions. @wrenoud tested some things in #3450 using `TEST_SET_LOCALE=nl_NL` in Windows.

> * There's some spots where creating big fat streams all the time are going to hurt performance. I tried to identify a few of those, but there might be some more. Updating them probably means changing PDAL's code a bit more.

I added @rouault's idea of a static thread_local stream to avoid the creation of the stream in `fromString` and a few more places that you pointed as critical.
Most of the changes tried to be minimal to not produce a big PR. But for performance reasons some parts of the code could be optimized as you say. I do not know which parts of the code are called massively, that could impact more the global performance.

jjimenezshaw commented:
Thanks @hobu I hope I addressed all your points.

> * There's not enough tests that have been added. ClassicLocaleStringStream and friends should have some explicit tests that demonstrate behavior and usage. Also, the place where locale stuff is the most difficult to deal with, Windows, is not tested at all.
> 

I added tests for ClassicLocale stringstreams. About Windows, unfortunately I do not have a computer to test, and neither the knowledge to configure the GitHub actions. @wrenoud tested some things in #3450 using `TEST_SET_LOCALE=nl_NL` in Windows.

> * There's some spots where creating big fat streams all the time are going to hurt performance. I tried to identify a few of those, but there might be some more. Updating them probably means changing PDAL's code a bit more.

I added @rouault's idea of a static thread_local stream to avoid the creation of the stream in `fromString` and a few more places that you pointed as critical.
Most of the changes tried to be minimal to not produce a big PR. But for performance reasons some parts of the code could be optimized as you say. I do not know which parts of the code are called massively, that could impact more the global performance.
